src/main/java/org/jabref/logic/cleanup/MoveFilesCleanup.java

```diff
@@ -16,64 +23,113 @@
 import org.jabref.model.entry.ParsedFileField;
 import org.jabref.model.metadata.FileDirectoryPreferences;
 
+import org.apache.commons.logging.Log;
+import org.apache.commons.logging.LogFactory;
+
 public class MoveFilesCleanup implements CleanupJob {
 
     private final BibDatabaseContext databaseContext;
     private final FileDirectoryPreferences fileDirectoryPreferences;
+    private final LayoutFormatterPreferences layoutPrefs;
+
+    private final String fileDirPattern;
+    private static final Log LOGGER = LogFactory.getLog(MoveFilesCleanup.class);
 
+    private ParsedFileField singleFileFieldCleanup;
 
-    public MoveFilesCleanup(BibDatabaseContext databaseContext, FileDirectoryPreferences fileDirectoryPreferences) {
+    public MoveFilesCleanup(BibDatabaseContext databaseContext, String fileDirPattern,
+            FileDirectoryPreferences fileDirectoryPreferences, LayoutFormatterPreferences layoutPrefs) {
         this.databaseContext = Objects.requireNonNull(databaseContext);
+        this.fileDirPattern = Objects.requireNonNull(fileDirPattern);
         this.fileDirectoryPreferences = Objects.requireNonNull(fileDirectoryPreferences);
+        this.layoutPrefs = Objects.requireNonNull(layoutPrefs);
+    }
+
+    public MoveFilesCleanup(BibDatabaseContext databaseContext, String fileDirPattern,
+            FileDirectoryPreferences fileDirectoryPreferences, LayoutFormatterPreferences prefs,
+            ParsedFileField field) {
+
+        this(databaseContext, fileDirPattern, fileDirectoryPreferences, prefs);
+        this.singleFileFieldCleanup = field;
     }
 
     @Override
     public List<FieldChange> cleanup(BibEntry entry) {
-        if(!databaseContext.getMetaData().getDefaultFileDirectory().isPresent()) {
+        Optional<Path> firstExistingFileDir = databaseContext.getFirstExistingFileDir(fileDirectoryPreferences);
```

Comment:
I don't like that so much logic is hidden in a cleanup process. In my opinion it would be better to extract as much logic as possible to separate classes and reuse them here and also in the other places (instead of reusing the cleanup).
In more detail I propose to:
- Create a class that wraps around `List<ParsedFileField>` and that takes care of moving all linked files (and providing nice helper methods like `replace(one file, by another one)` 
- Almost everything in the [for loop below](https://github.com/JabRef/jabref/pull/2415/files#diff-697281a5461cd043769c3bbf1331a9deR87) can be extracted to the ParsedFileField class (and be split in smaller helper methods like ParsedFileField.exists())
- The code is a bit complicated since you constantly have to pass the fileDirectoryPreferences around (not your fault!). What you think about passing these information once into `TypedBibEntry.getFiles` as arguments, which are then stored in the ParsedFileField class? Thus the ParsedFileField class actually knowns if the file exists in one of the directories, what its preferred name is, can rename and move itself...